<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using create statements and referencing local data frames • gorr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using create statements and referencing local data frames">
<meta property="og:description" content="gorr">
<meta property="og:image" content="/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gorr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basic-query.html">Executing a Basic GOR Query</a>
    </li>
    <li>
      <a href="../articles/gor-create.html">Using create statements and referencing local data frames</a>
    </li>
    <li>
      <a href="../articles/persistance-of-results.html">Server-side Persistance of Query Results</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/wuxi-nextcode/gorr/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="gor-create_files/header-attrs-2.3/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using create statements and referencing local data frames</h1>
                        <h4 class="author">Edvald Gislason <a href="mailto:edvald@wuxinextcode.com" class="email">edvald@wuxinextcode.com</a>
</h4>
            
            <h4 class="date">2020-10-07</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/wuxi-nextcode/gorr/blob/master/vignettes/gor-create.Rmd"><code>vignettes/gor-create.Rmd</code></a></small>
      <div class="hidden name"><code>gor-create.Rmd</code></div>

    </div>

    
    
<p>In this example, we’re going to show how the <code>gor_create</code> can be used to prepare and construct a query <a href="https://en.wikipedia.org/wiki/Closure_(computer_programming)">closure</a>. This both reduces repetitions in code, as well as simplifies iterative workflows in GOR.</p>
<div id="load-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#load-packages" class="anchor"></a>Load packages</h2>
<p>First load the <code>gorr</code> package, the <code>tidyverse</code> package is recommended, but for the sake of simplicity we pick out the ones we’re using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wuxi-nextcode/gorr">gorr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">magrittr</span><span class="op">)</span> <span class="co"># pipe</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span></pre></div>
<p>Next we make a <code>conn</code> object for holding information on the API we’re connecting to. <code>gor_connect</code> takes 2 parameters <code>api_key</code> and <code>project</code> but if either are left out then it will try to read the environment variables <code>GOR_API_KEY</code>, and <code>GOR_API_PROJECT</code> respectively. Here below we have the <code>GOR_API_KEY</code> environment variable already defined so supplying the function only with a target project suffices. After this we create a <code>query</code> function/closure so we don’t have to reference <code>conn</code> again:</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="va">conn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gor_connect.html">gor_connect</a></span><span class="op">(</span>project <span class="op">=</span> <span class="st">"ukbb_hg38"</span><span class="op">)</span>
<span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gor_create.html">gor_create</a></span><span class="op">(</span>conn <span class="op">=</span> <span class="va">conn</span><span class="op">)</span>
<span class="va">query</span>
<span class="co">#&gt; ── GOR Creation Query ──────────────────────────────────────────────────────────</span>
<span class="co">#&gt; Connection</span>
<span class="co">#&gt;  Service Root: https://staging.wuxinextcodedev.com/api/query</span>
<span class="co">#&gt;  Project: ukbb_hg38</span>
<span class="co">#&gt; Definitions</span>
<span class="co">#&gt;   None</span>
<span class="co">#&gt; Create statements &amp; virtual relations</span>
<span class="co">#&gt;   None</span></pre></div>
<p>Now we can call that function with only the query parameter. Let’s search for genes containing <em>BRCA</em> and save the resulting table as a local dataframe <code>mygenes</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">mygenes</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="st">"gor #genes# | grep BRCA"</span><span class="op">)</span>
<span class="va">mygenes</span>
<span class="co">#&gt; # A tibble: 2 x 4</span>
<span class="co">#&gt;   chrom gene_start gene_end gene_symbol</span>
<span class="co">#&gt;   &lt;chr&gt;      &lt;int&gt;    &lt;int&gt; &lt;chr&gt;      </span>
<span class="co">#&gt; 1 chr13   32315085 32400266 BRCA2      </span>
<span class="co">#&gt; 2 chr17   43044294 43170245 BRCA1</span></pre></div>
<p>Next we can expand on our previously defined <code>query</code> function by supplying it back into <code>gor_create</code> as the <code>replace</code> parameter. This time we include some definitions using the <code>defs</code> parameter and then we can alias our local dataframe so that we can reference it in remote queries. In GOR this is called <em>virtual relations</em>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gor_create.html">gor_create</a></span><span class="op">(</span>
  defs <span class="op">=</span> <span class="st">"def variants = #dbsnp#"</span>,
  mygenes <span class="op">=</span> <span class="va">mygenes</span>, 
  replace <span class="op">=</span> <span class="va">query</span>
<span class="op">)</span>
<span class="va">query</span>
<span class="co">#&gt; ── GOR Creation Query ──────────────────────────────────────────────────────────</span>
<span class="co">#&gt; Connection</span>
<span class="co">#&gt;  Service Root: https://staging.wuxinextcodedev.com/api/query</span>
<span class="co">#&gt;  Project: ukbb_hg38</span>
<span class="co">#&gt; Definitions</span>
<span class="co">#&gt;   def variants = #dbsnp#;</span>
<span class="co">#&gt; Create statements &amp; virtual relations</span>
<span class="co">#&gt;  mygenes</span>
<span class="co">#&gt; Warning: `tbl_df()` is deprecated as of dplyr 1.0.0.</span>
<span class="co">#&gt; Please use `tibble::as_tibble()` instead.</span>
<span class="co">#&gt; This warning is displayed once every 8 hours.</span>
<span class="co">#&gt; Call `lifecycle::last_warnings()` to see where this warning was generated.</span>
<span class="co">#&gt;    # A tibble: 2 x 4</span>
<span class="co">#&gt;      chrom gene_start gene_end gene_symbol</span>
<span class="co">#&gt;      &lt;chr&gt;      &lt;int&gt;    &lt;int&gt; &lt;chr&gt;      </span>
<span class="co">#&gt;    1 chr13   32315085 32400266 BRCA2      </span>
<span class="co">#&gt;    2 chr17   43044294 43170245 BRCA1</span></pre></div>
<p>Now that we have our updated <code>query</code> function, we can use it to <code>gor</code> our table of genes and join it to the <code>#dbsnp#</code> table we aliased as <code>variants</code> in the definitions part above. The result is a list of all variants within each gene in our table</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="va">brca_variants</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="st">"
    gor [mygenes] | join -segvar variants        
"</span><span class="op">)</span>

<span class="va">brca_variants</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">gene_symbol</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>records <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span>,
            variants <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/n_distinct.html">n_distinct</a></span><span class="op">(</span><span class="va">rsids</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; `summarise()` ungrouping output (override with `.groups` argument)</span>
<span class="co">#&gt; # A tibble: 2 x 3</span>
<span class="co">#&gt;   gene_symbol records variants</span>
<span class="co">#&gt;   &lt;chr&gt;         &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1 BRCA1         37150    32933</span>
<span class="co">#&gt; 2 BRCA2         28727    25260</span></pre></div>
<p>The reason for the difference in <code># records</code>and <code># variants</code> above can be explained by looking into the data:</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="va">target_variant</span> <span class="op">&lt;-</span> 
  <span class="va">brca_variants</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">rsids</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">rsids</span><span class="op">)</span>

<span class="va">target_variant</span>
<span class="co">#&gt; [1] "rs71071031"</span></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="va">brca_variants</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">rsids</span> <span class="op">==</span> <span class="va">target_variant</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">distance</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 23 x 8</span>
<span class="co">#&gt;    chrom gene_start gene_end gene_symbol     pos reference        allele rsids  </span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;int&gt;    &lt;int&gt; &lt;chr&gt;         &lt;int&gt; &lt;chr&gt;            &lt;chr&gt;  &lt;chr&gt;  </span>
<span class="co">#&gt;  1 chr13   32315085 32400266 BRCA2        3.24e7 TTTTTTTTTTTTTTT… T      rs7107…</span>
<span class="co">#&gt;  2 chr13   32315085 32400266 BRCA2        3.24e7 TTTTTTTTTTTTTTT… T      rs7107…</span>
<span class="co">#&gt;  3 chr13   32315085 32400266 BRCA2        3.24e7 TTTTTTTTTTTTTTT… T      rs7107…</span>
<span class="co">#&gt;  4 chr13   32315085 32400266 BRCA2        3.24e7 TTTTTTTTTTTTTTT… T      rs7107…</span>
<span class="co">#&gt;  5 chr13   32315085 32400266 BRCA2        3.24e7 TTTTTTTTTTTTTTT… T      rs7107…</span>
<span class="co">#&gt;  6 chr13   32315085 32400266 BRCA2        3.24e7 TTTTTTTTTTTTTTT… T      rs7107…</span>
<span class="co">#&gt;  7 chr13   32315085 32400266 BRCA2        3.24e7 TTTTTTTTTTTTTTT… T      rs7107…</span>
<span class="co">#&gt;  8 chr13   32315085 32400266 BRCA2        3.24e7 TTTTTTTTTTTTTTT… T      rs7107…</span>
<span class="co">#&gt;  9 chr13   32315085 32400266 BRCA2        3.24e7 TTTTTTTTTTTTTTT… T      rs7107…</span>
<span class="co">#&gt; 10 chr13   32315085 32400266 BRCA2        3.24e7 TTTTTTTTTTTTTTTT T      rs7107…</span>
<span class="co">#&gt; # … with 13 more rows</span></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Genuity Science, Edvald Gislason.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
